### ISSUE TITLE
Nested generic models fail to validate when using 'Optional' at toplevel

### ISSUE THREAD

**Issue by nu11ptr** [samuelcolvin/pydantic#2331]
### Checks

* [X] I added a descriptive title to this issue
* [X] I have searched (google, github) for similar issues and couldn't find anything
* [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.7.3
            pydantic compiled: True
                 install path: /home/scott/src/network_api_v2/.env/lib/python3.8/site-packages/pydantic
               python version: 3.8.5 (default, Jul 28 2020, 12:59:40)  [GCC 9.3.0]
                     platform: Linux-5.4.0-47-generic-x86_64-with-glibc2.29
     optional deps. installed: ['typing-extensions']
```

```py
from ipaddress import IPv4Network
from typing import Generic, Optional, TypeVar

from pydantic.generics import GenericModel
from pydantic import BaseModel

T = TypeVar("T")

class Child(GenericModel, Generic[T]):
    field: T

class Parent(GenericModel, Generic[T]):
    child: Optional[Child[T]] = None

class Net(BaseModel):
    ipnet: IPv4Network
```

Using this with invalid data in `Net` results in:

```py
>>> Parent[Net].parse_obj({"child": {"field": {"ipnet": "1.1.1/24"}}})
Parent[Net](child=Child[T](field={'ipnet': '1.1.1/24'}))
```

Redefining `Parent` minus `Optional` fixes it:

```py
>>> class Parent(GenericModel, Generic[T]):
...     child: Child[T] = None
... 
>>> Parent[Net].parse_obj({"child": {"field": {"ipnet": "1.1.1/24"}}})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 520, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 362, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Parent[Net]
child -> field -> ipnet
  value is not a valid IPv4 network (type=value_error.ipv4network)
```

**Comment by PrettyWood** [comment]
Hi @nu11ptr 
Your issue has already been solved by PR #1989.
The fix will be available in pydantic 1.8 for sure 👍
In the meantime you can always use the merge commit and install pydantic directly from the repository
`pip install git+https://github.com/samuelcolvin/pydantic.git@6b8d2babaaca420d20aab52bfe5f42cf5e382f1d`
Hope it helps :)

**Comment by nu11ptr** [comment]
Ahh...cool. Didn't see that one. Thanks for the quick response!
